muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/rest/CommunicatorRecipientsRESTService.java

```diff
@@ -0,0 +1,308 @@
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.TEACHER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_GUIDER);
+    } else {
+      // Default for other roles
+      
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.ADMINISTRATOR);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.MANAGER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_PROGRAMME_LEADER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.TEACHER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDY_GUIDER);
+      roleArchetypeFilter.add(EnvironmentRoleArchetype.STUDENT);
+      
+      if (sessionController.hasEnvironmentPermission(RoleFeatures.ACCESS_ONLY_GROUP_STUDENTS)) {
+        // Study guider limitations - groups where user is a member
+        List<UserGroupEntity> userGroups = userGroupEntityController.listUserGroupsByUserIdentifier(sessionController.getLoggedUser());
+        userGroupFilters = userGroups.stream().map(userGroup -> userGroup.getId()).collect(Collectors.toSet());
+      }
+    }
+
+    SearchProvider searchProvider = getSearchProvider();
+    if (searchProvider != null) {
+      String[] fields = new String[] { "firstName", "lastName", "nickName", "email" };
+
+      Boolean onlyDefaultUsers = true;
+      
+      SearchResult result = searchProvider.searchUsers(searchString, 
+          fields, 
+          roleArchetypeFilter, 
+          userGroupFilters, 
+          workspaceFilters, 
+          null, 
+          false,
+          false,
+          onlyDefaultUsers,
+          firstResult, 
+          maxResults);
+      
+      List<Map<String, Object>> results = result.getResults();
+
+      List<fi.otavanopisto.muikku.rest.model.User> ret = new ArrayList<fi.otavanopisto.muikku.rest.model.User>();
+
+      for (Map<String, Object> o : results) {
+        String[] id = ((String) o.get("id")).split("/", 2);
+        UserEntity userEntity = userEntityController
+            .findUserEntityByDataSourceAndIdentifier(id[1], id[0]);
+        
+        if (userEntity != null) {
+          boolean hasImage = userEntityFileController.hasProfilePicture(userEntity);
+          String emailAddress = userEmailEntityController.getUserDefaultEmailAddress(userEntity, true);
+          Date studyStartDate = getDateResult(o.get("studyStartDate"));
+          Date studyTimeEnd = getDateResult(o.get("studyTimeEnd"));
+
+          ret.add(new fi.otavanopisto.muikku.rest.model.User(
+            userEntity.getId(), 
+            (String) o.get("firstName"),
+            (String) o.get("lastName"), 
+            (String) o.get("nickName"), 
+            hasImage,
+            (String) o.get("nationality"),
+            (String) o.get("language"), 
+            (String) o.get("municipality"), 
+            (String) o.get("school"), 
+            emailAddress,
+            studyStartDate,
+            studyTimeEnd));
+        }
```

Comment:
Worth logging a warning in else block? Users found in search index but not in database is something that shouldn't really happen (but sometimes does, especially when fooling around with dev environment).